base/channels.jl

```diff
@@ -318,15 +319,19 @@ end
 
 function put_buffered(c::Channel, v)
     lock(c)
+    did_buffer = false
     try
+        @atomic :monotonic c.length += 1
         while length(c.data) == c.sz_max
             check_channel_state(c)
             wait(c.cond_put)
         end
         push!(c.data, v)
+        did_buffer = true
```

Comment:
I am slightly confused what this is a count of precisely. Why not do the increment here, so it is always a lower bound on the number of items which are available, rather than being slightly ahead of the items that are available?

Reply:
This allows us to count the tasks waiting in `wait(c.cond_put)`, for consistency with unbuffered channels.

If we don't do this, the count of "available items" is quite inconsistent between (finite) buffered vs unbuffered channels.

Reply:
By the way, I don't have a strong opinion on the "right" answer here. I just think we should be consistent in the way we count "available" items in buffered/unbuffered cases.

Reply:
Yeah, I agree it's nice to be consistent.

Another approach to make it consistent is to have `length(unbuffered_channel) == 0` and `length(buffered_channel) == length(buffered_channel.data)`, right? Maybe it's reasonable in some sense, since we'd have

```julia
n = length(ch)
close(ch)
@assert length(collect(ch)) == n
```

when there are no takers. This property seems intuitive to me. (Though I guess it makes `length` rather useless for the internals...)

Reply:
Yes this is definitely the other option :+1:

I think the current system might be more useful as it gives you more options for applying backpressure?

Reply:
Maybe the simplest approach would be to keep using the name `n_avail` instead of `length`?

Reply:
ie, if `length(unbuffered_channel) == 0`, there's no real way to tell whether an upstream provider should slow down if it fires off a task to generate each item.

Reply:
Yeah, the current implementation is definitely more informative since what I proposed can be implemented as `min(sz_max, length(ch))` but not the other way around. I was just wondering what `length` even means here...

Though we can also explain the "invariance" of the current implementation by

```julia
lock(ch)
for _ in 1:length(ch)
    take!(ch)  # does not block (or maybe rather "eventually succeeds")
end
unlock(ch)
```

which seems to be a nice property.

Reply:
> I was just wondering what length even means here.

I'm thinking of the channel as a container where `length()` means the number of items "in" the container; the number of times you should be able to `take!()` and still make progress in a single-taker situation.

Is this an acceptable way to use `length()`? For concurrent containers this kind of thing should always be viewed with some suspicion, I guess :-)

The other thing I wonder is whether `length()` should be considered an `O(1)` operation that we want to expose for channels in the long term. Say we changed the implementation to one of the more fancy lock-free queues, would it still make sense to have `length()`? (Though I'm not sure this can be done anyway, as it would break API compatibility for `lock(c)`?)

Reply:
> reverted to `n_avail`

Ah, sorry, I didn't notice.

> the former condition, in a batch processing system, is relevant for flow-control of writers, and the later is relevant for flow-control of readers

Hmm... interesting. This feels like an important point. Implementing this idea requires supporting `length(c.data)` and `n_avail` separately.